pulp_python/tests/functional/api/test_pypi_apis.py

```diff
@@ -64,6 +76,182 @@ def test_published_index(self):
+            files={"content": open(self.egg, "rb")},
+        )
+        response2 = session.post(
+            url,
+            data={"sha256_digest": PYTHON_WHEEL_SHA256},
+            files={"content": open(self.wheel, "rb")},
+        )
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response2.status_code, 200)
+        response, response2 = response.json(), response2.json()
+        self.assertEqual(response["session"], response2["session"])
+        self.assertEqual(response2["task"], "updated")
+        monitor_task(response["task"])
+        content = get_added_content_summary(repo, f"{repo.versions_href}1/")
+        self.assertDictEqual({PYTHON_CONTENT_NAME: 2}, content)
+
+    def test_package_upload_session_long(self):
+        """Tests that long uploads will be broken up into multiple tasks."""
+        repo, distro = self._create_empty_repo_and_distribution()
+        url = urljoin(PYPI_HOST, distro.base_path + "/legacy/")
+        session = requests.Session()
+        response = session.post(
+            url,
+            data={"sha256_digest": PYTHON_EGG_SHA256},
+            files={"content": open(self.egg, "rb")},
+        )
+        self.assertEqual(response.status_code, 200)
+        response = response.json()
+        monitor_task(response["task"])
+        response2 = session.post(
+            url,
+            data={"sha256_digest": PYTHON_WHEEL_SHA256},
+            files={"content": open(self.wheel, "rb")},
+        )
+        self.assertEqual(response2.status_code, 200)
+        response2 = response2.json()
+        self.assertNotEqual("updated", response2["task"])
+        self.assertNotEqual(response["session"], response2["session"])
+        self.assertNotEqual(response["task"], response2["task"])
+        monitor_task(response2["task"])
+        content = get_content_summary(repo, f"{repo.versions_href}2/")
+        self.assertDictEqual({PYTHON_CONTENT_NAME: 2}, content)
+
+    def test_package_upload_simple(self):
+        """Tests that the package upload endpoint exposed at `/simple/` works."""
+        repo, distro = self._create_empty_repo_and_distribution()
+        url = urljoin(PYPI_HOST, distro.base_path + "/simple/")
+        response = requests.post(
+            url,
+            data={"sha256_digest": PYTHON_EGG_SHA256},
+            files={"content": open(self.egg, "rb")},
+        )
+        self.assertEqual(response.status_code, 200)
+        task = response.json()["task"]
+        monitor_task(task)
+        content = get_added_content_summary(repo, f"{repo.versions_href}1/")
+        self.assertDictEqual({PYTHON_CONTENT_NAME: 1}, content)
+
+    def test_twine_upload(self):
+        """Tests that packages can be properly uploaded through Twine."""
+        repo, distro = self._create_empty_repo_and_distribution()
+        url = urljoin(PYPI_HOST, distro.base_path + "/legacy/")
+        username, password = "admin", "password"
+        subprocess.run(
```

Comment:
It's been a long, long time since I touched this functionality, but I would check to see if `pulp_smash.cli.Client` provides any extra functionality that makes it better suited.

(This isn't a requirement, just a suggestion, I actually don't know that it does.  We don't often drive a shell from functional tests)

Reply:
Tried to use it originally, and it just wasn't well suited. The problem is that it runs the commands in the same container as the Pulp instance, but our CI installs the test_requirements in the VM running that container, so the pulp_smash CLI has no access to `twine`.

Reply:
+1